src/Npgsql/NpgsqlBatch.cs

```diff
@@ -171,4 +171,30 @@ public override Task PrepareAsync(CancellationToken cancellationToken = default)
 
     /// <inheritdoc />
     public override void Cancel() => Command.Cancel();
+
+    /// <inheritdoc />
+    public override void Dispose()
+    {
+        ClearTransaction();
+        if (Command.IsCached && Connection is not null && Connection.CachedBatch is null)
+        {
+            BatchCommands.Clear();
+            Connection.CachedBatch = this;
```

Comment:
I think there are other things that need to be reset - EnableErrorBarriers, AllResultTypesAreUnknown (on command too?), maybe do a pass to make sure.

Reply:
Good catch. Fixed that.